File: 200-number-of-islands/number-of-islands.cpp
class Solution {
    int direction[4][2]={{1,0},{-1,0},{0,1},{0,-1}};
public:
void dfs(vector<vector<char>>& grid, int r, int c)
{
    if(r>=grid.size()||c>=grid[0].size()||grid[r][c]=='0') return;

    grid[r][c]='0';
    for(int i=0;i<4;i++)
    {
        dfs(grid,r+direction[i][0],c+direction[i][1]);
    }
}
    int numIslands(vector<vector<char>>& grid) {
        int row=grid.size();
        int cols=grid[0].size();
        int islands=0;

        for(int r=0;r<row;r++)
        {
            for(int c=0;c<cols;c++)
            {
                if(grid[r][c]=='1')
                {
                    dfs(grid,r,c);
                    islands++;
                }
            }
        }

        return islands;
        
    }
};

File: 200-number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:


        q=deque()
        islands=0

        for r in range(len(grid)):
            for c in range(len(grid[0])):
                if grid[r][c]=='1':
                    islands+=1
                    q.append([r,c])
                    grid[r][c]='0'

                    while q:

                        row,col=q.popleft()

                        # grid[row][col]='0'

                        for dr,dc in [(1,0),(-1,0),(0,1),(0,-1)]:
                            nr=row+dr
                            nc=col+dc

                            if 0<=nr<len(grid) and 0<=nc<len(grid[0]) and grid[nr][nc]=='1':
                                grid[nr][nc]='0'
                                q.append([nr,nc])
        return islands




        